llvm/lib/IR/Instruction.cpp

```diff
@@ -975,6 +975,14 @@ bool Instruction::isIdenticalToWhenDefined(const Instruction *I,
     return equal(Phi->blocks(), OtherPhi->blocks());
   }
 
+  if (const SwitchInst *SI = dyn_cast<SwitchInst>(this)) {
+    const SwitchInst *OtherSI = cast<SwitchInst>(I);
+    for (auto [Case, OtherCase] : zip(SI->cases(), OtherSI->cases()))
+      if (Case.getCaseValue() != OtherCase.getCaseValue())
+        return false;
+    return true;
+  }
```

Comment:
I'd expect this check to be inside hasSameSpecialState(). And then hopefully the SimplifyCFG change is not needed.

Reply:
This probably also needs a change to FunctionComparator (testable via MergeFunctions).